service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,156 @@
+import com.epam.esm.entities.GiftCertificate;
+import com.epam.esm.entities.Tag;
+import com.epam.esm.exception.InvalidFieldException;
+import com.epam.esm.exception.ResourceNotFoundException;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.validator.GiftCertificateValidator;
+import com.epam.esm.mapper.GiftCertificateDtoMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.*;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService<GiftCertificate> {
+
+    private final GiftCertificateDao giftCertificateDao;
+    private final TagDao tagDao;
+    private final GiftCertificateTagDao giftCertificateTagDao;
+    private final GiftCertificateValidator giftCertificateValidator;
+    private final GiftCertificateDtoMapper giftCertificateDtoWrapper;
+    private final String TIME_ZONE = "GMT+3";
+
+
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, TagDao tagDao,
+                                      GiftCertificateTagDao giftCertificateTagDao,
+                                      GiftCertificateValidator giftCertificateValidator,
+                                      GiftCertificateDtoMapper giftCertificateDtoMapper) {
+        this.giftCertificateDao = giftCertificateDao;
+        this.tagDao = tagDao;
+        this.giftCertificateTagDao = giftCertificateTagDao;
+        this.giftCertificateValidator = giftCertificateValidator;
+        this.giftCertificateDtoWrapper = giftCertificateDtoMapper;
+    }
+
+
+    @Override
+    public List<GiftCertificateDto> getGiftCertificates(String tagName, String giftCertificateName, String description,
+                                                        String sortByName, String sortByDate) {
+        List<GiftCertificate> sortedGiftCertificates = giftCertificateDao.getGiftCertificates(tagName,
+                giftCertificateName, description, sortByName, sortByDate);
+
+        for (GiftCertificate giftCertificate : sortedGiftCertificates){
+            Set <Tag> tags = getTagsByGiftCertificateId(giftCertificate.getCertificateId());
+            giftCertificate.setTags(tags);
+        }
+
+        List<GiftCertificateDto> sortedDtoGiftCertificates = new ArrayList<>();
+
+        sortedGiftCertificates.forEach(giftCertificate ->
+                sortedDtoGiftCertificates.add(giftCertificateDtoWrapper.map(giftCertificate)));
+        return sortedDtoGiftCertificates;
+    }
+
+    @Override
+    public GiftCertificateDto getGiftCertificate(long id) throws ResourceNotFoundException {
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findById(id);
+        if (optionalGiftCertificate.isEmpty()) {
+            throw new ResourceNotFoundException(id);
+        }
```

Comment:
GiftCertificate giftCertificate = giftCertificateDao.findById(id).orElseThrow(() -> new ResourceNotFoundException(id)) instead of 'if'

Reply:
Refactor this code